pep-0705.rst

```diff
@@ -0,0 +1,298 @@
+
+    def f(a: A) -> None:
+        print(a.get('x', 0))
+
+    b: B = {'x': 0}
+    f(b)  # Accepted by type checker
+
+* A ``TypedMapping`` type ``A`` with no key ``'x'`` is not consistent with a ``TypedMapping`` type with a non-required key ``'x'``, since at runtime the key ``'x'`` could be present and have an incompatible type (which may not be visible through ``A`` due to structural subtyping). This is the same as for ``TypedDict`` types. Example::
+
+    class A(TypedMapping, total=False):
+        x: int
+        y: int
+
+    class B(TypedMapping, total=False):
+        x: int
+
+    class C(TypedMapping, total=False):
+        x: int
+        y: str
+
+    def f(a: A) -> None:
+        print(a.get('y') + 1)
+
+    def g(b: B) -> None:
+        f(b)  # Type check error: 'B' incompatible with 'A'
+
+    c: C = {'x': 0, 'y': 'foo'}
+    g(c)  # Runtime error: str + int
+
+* A ``TypedMapping`` with all ``int`` values is not consistent with ``Mapping[str, int]``, since there may be additional non-``int`` values not visible through the type, due to structural subtyping. This mirrors ``TypedDict``. Example::
+
+    class A(TypedMapping):
+        x: int
+
+    class B(TypedMapping):
+        x: int
+        y: str
+
+    def sum_values(m: Mapping[str, int]) -> int:
+        return sum(m.values())
+
+    def f(a: A) -> None:
+        sum_values(a)  # Type check error: 'A' incompatible with Mapping[str, int]
+
+    b: B = {'x': 0, 'y': 'foo'}
+    f(b)  # Runtime error: int + str
+
+
+Backwards Compatibility
+=======================
+
+This PEP changes the rules for how ``TypedDict`` behaves (allowing subclasses to
+inherit from ``TypedMapping`` protocols in a way that changes the resulting
+overloads), so code that inspects ``TypedDict`` types will have to change. This
+is expected to mainly affect type-checkers.
+
+The ``TypedMapping`` type will be added to the ``typing_extensions`` module,
+enabling its use in older versions of Python. Users wanting to inherit from
+both ``TypedDict`` and ``TypedMapping`` will need to use
+``typing_extensions.TypedDict`` in these versions, as the existing ``typing`` implementation
+of ``TypedDict`` does an ``is`` check on the metaclass of all base classes and raises
+an error if any other metaclass is found. This would cause any subclass of both
+that implementation and a ``TypedMapping`` protocol to trigger an error at runtime.
+
+
+Security Implications
+=====================
+
+There are no known security consequences arising from this PEP.
+
+
+How to Teach This
+=================
+
+This PEP should be linked in the :mod:`typing` module's documentation.
+Class documentation should also be added, using that for
+:class:`~collections.abc.Mapping`, :class:`~typing.Protocol` and
+:class:`~typing.TypedDict` as examples.
+
+Suggested introductory sentence: "Base class for read-only mapping protocol classes."
+
+
+Reference Implementation
+========================
+
+No reference implementation exists yet.
```

Comment:
I would encourage you to submit a typing-extensions implementation as soon as possible.

Reply:
Understood, thanks.